File: native/src/spatial_mapping.cpp
#include "spatial_mapping.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <vector>

namespace pa {

double extract_amplitude_at_frequency(const double* freqs, const double* magnitude,
                                      std::size_t n, double target_freq_hz,
                                      double tolerance_hz) {
    if (n == 0 || freqs == nullptr || magnitude == nullptr) return 0.0;
    double best = -std::numeric_limits<double>::infinity();
    bool any = false;
    for (std::size_t i = 0; i < n; ++i) {
        if (std::abs(freqs[i] - target_freq_hz) <= tolerance_hz) {
            any = true;
            best = std::max(best, magnitude[i]);
        }
    }
    return any ? best : 0.0;
}

double estimasi_noise_floor(const double* freqs, const double* mag, std::size_t n,
                            double target_freq_hz, double tolerance_hz,
                            double sideband_factor) {
    if (n == 0 || freqs == nullptr || mag == nullptr) return 0.0;
    std::vector<double> side;
    side.reserve(n);
    for (std::size_t i = 0; i < n; ++i) {
        const double jarak = std::abs(freqs[i] - target_freq_hz);
        if (jarak > tolerance_hz && jarak <= sideband_factor * tolerance_hz) {
            side.push_back(mag[i]);
        }
    }
    if (side.empty()) return 0.0;
    std::nth_element(side.begin(), side.begin() + side.size() / 2, side.end());
    return side[side.size() / 2];
}

GrayScaleResult amplitude_matrix_to_grayscale(
    const double* matrix, int rows, int cols,
    const unsigned char* captured_mask,
    bool has_fixed_range, double amp_min_fixed, double amp_max_fixed) {
    GrayScaleResult out;
    out.rows = rows;
    out.cols = cols;
    const std::size_t n = static_cast<std::size_t>(rows) * static_cast<std::size_t>(cols);
    out.gray.assign(n, 0);

    if (rows <= 0 || cols <= 0 || matrix == nullptr) {
        return out;
    }

    double amp_min = 0.0;
    double amp_max = 0.0;

    if (has_fixed_range) {
        amp_min = amp_min_fixed;
        amp_max = amp_max_fixed;
    } else {
        bool any = false;
        amp_min = std::numeric_limits<double>::infinity();
        amp_max = -std::numeric_limits<double>::infinity();
        for (std::size_t i = 0; i < n; ++i) {
            if (captured_mask != nullptr && captured_mask[i] == 0) continue;
            any = true;
            amp_min = std::min(amp_min, matrix[i]);
            amp_max = std::max(amp_max, matrix[i]);
        }
        if (!any) {
            out.amp_min = 0.0;
            out.amp_max = 0.0;
            return out;
        }
    }

    out.amp_min = amp_min;
    out.amp_max = amp_max;

    if (amp_max <= amp_min) {
        std::fill(out.gray.begin(), out.gray.end(), static_cast<unsigned char>(128));
        return out;
    }

    for (std::size_t i = 0; i < n; ++i) {
        double normalized = (matrix[i] - amp_min) / (amp_max - amp_min);
        if (normalized < 0.0) normalized = 0.0;
        if (normalized > 1.0) normalized = 1.0;
        const int v = static_cast<int>(normalized * 255.0 + 0.5);
        out.gray[i] = static_cast<unsigned char>(std::clamp(v, 0, 255));
    }
    return out;
}

}  // namespace pa

File: backend/spatial_mapping.py
"""
Ekstraksi amplitudo pada frekuensi target + normalisasi grayscale.

Aturan citra (Set Modulasi = f_set, mis. 17000 Hz):
  - f < f_set  → background hitam (tidak dipakai sebagai objek)
  - f ≈ f_set  → objek (amplitudo di jendela toleransi)

Pemetaan grayscale:
  - amplitudo tinggi → terang
  - amplitudo rendah → gelap
"""

import numpy as np


def extract_amplitude_at_frequency(freqs, magnitude, target_freq_hz,
                                   tolerance_hz=50.0):
    freqs = np.asarray(freqs)
    magnitude = np.asarray(magnitude)
    if len(freqs) == 0:
        return 0.0

    mask = np.abs(freqs - target_freq_hz) <= tolerance_hz
    if not mask.any():
        return 0.0
    return float(magnitude[mask].max())


def extract_amplitude_object_black_background(
    freqs, magnitude, target_freq_hz, tolerance_hz=50.0
):
    """
    Amplitudo objek dengan pita di bawah frekuensi set sebagai background hitam.

    Spektrum pada f < target_freq_hz di-nol-kan (background hitam / plat),
    lalu amplitudo diambil hanya di sekitar target_freq_hz (± tolerance_hz).
    """
    freqs = np.asarray(freqs)
    magnitude = np.asarray(magnitude, dtype=np.float64)
    if len(freqs) == 0:
        return 0.0

    mag = magnitude.copy()
    mag[freqs < float(target_freq_hz)] = 0.0
    return extract_amplitude_at_frequency(
        freqs, mag, target_freq_hz, tolerance_hz
    )


def estimasi_noise_floor(freqs, mag, target_freq_hz, tolerance_hz,
                         sideband_factor=5.0):
    jarak = np.abs(np.asarray(freqs) - target_freq_hz)
    mask = (jarak > tolerance_hz) & (jarak <= sideband_factor * tolerance_hz)
    # Sideband di bawah f_set sudah dianggap background; pakai sisi atas saja
    # agar lantai derau tidak terkontaminasi energi plat.
    mask = mask & (np.asarray(freqs) >= float(target_freq_hz))
    if not mask.any():
        return 0.0
    return float(np.median(np.asarray(mag)[mask]))


def amplitude_matrix_to_grayscale(
    matrix,
    captured_mask=None,
    amp_min_fixed=None,
    amp_max_fixed=None,
    invert=False,
):
    """
    Normalisasi matrix amplitudo ke grayscale 0..255.

    invert=False (default): amp tinggi → terang, amp rendah → gelap.
    invert=True: amp rendah → terang, amp tinggi → gelap
    (cocok sample lebih lembek dari plat → amp turun).
    """
    matrix = np.asarray(matrix, dtype=np.float64)

    if amp_min_fixed is not None and amp_max_fixed is not None:
        amp_min, amp_max = float(amp_min_fixed), float(amp_max_fixed)
    else:
        if captured_mask is not None:
            mask = np.asarray(captured_mask, dtype=bool)
            nilai_valid = matrix[mask] if mask.any() else matrix.flatten()
        else:
            nilai_valid = matrix.flatten()

        if nilai_valid.size == 0:
            return np.zeros(matrix.shape, dtype=np.uint8), 0.0, 0.0

        amp_min = float(nilai_valid.min())
        amp_max = float(nilai_valid.max())

    if amp_max <= amp_min:
        # Semua sama: jika nol → hitam (background); jika tidak → abu-abu
        if amp_max <= 0.0:
            grayscale = np.zeros(matrix.shape, dtype=np.uint8)
        else:
            grayscale = np.full(matrix.shape, 128, dtype=np.uint8)
        return grayscale, amp_min, amp_max

    normalized = (matrix - amp_min) / (amp_max - amp_min)
    normalized = np.clip(normalized, 0.0, 1.0)
    if invert:
        # amp rendah (sample lembek) → 255 terang; amp tinggi (plat) → 0 gelap
        normalized = 1.0 - normalized
    grayscale = np.clip(np.round(normalized * 255.0), 0, 255).astype(np.uint8)
    return grayscale, amp_min, amp_max
